### PR TITLE
Rule builder

### LINKED ISSUE #2865: Rule Builder implementation for Angular
https://carbon-addons-iot-react.com/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-rulebuilder--rule-builder

https://pages.github.ibm.com/ai-applications/design/experimental/rulebuilder/v2

### PR DESCRIPTION
**Pull request by zvonimirfras**
Closes #2865

### CODE REVIEW

File: packages/angular/src/rule-builder/rule-builder-header.component.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  template: `
+    <ai-rule-builder-group-logic
+      [id]="id"
+      [selected]="groupLogic"
+      (selectedChange)="groupLogicChange.emit($event)"
+    >
+    </ai-rule-builder-group-logic>
+    <div class="iot--rule-builder-header__buttons">
+      <button ibmButton="ghost" (click)="addRule.emit({})">
+        Add rule
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-header.component.ts:14]: Should this text be modifiable?

File: packages/angular/src/rule-builder/rule-builder-header.component.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        <svg class="bx--btn__icon" ibmIcon="add" size="32"></svg>
+        <span class="bx--assistive-text">Add new rule</span>
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-header.component.ts:16]: Should this text be modifiable?

File: packages/angular/src/rule-builder/rule-builder-header.component.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      </button>
+      <button ibmButton="ghost" (click)="addRule.emit({ isGroup: true })">
+        Add group
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-header.component.ts:19]: Should this text be modifiable?

File: packages/angular/src/rule-builder/rule-builder-header.component.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        <svg class="bx--btn__icon" ibmIcon="text--new-line" size="32"></svg>
+        <span class="bx--assistive-text">Add new rule group</span>
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-header.component.ts:21]: Should this text be modifiable?

File: packages/angular/src/rule-builder/rule-builder-group-logic.component.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+@Component({
+  selector: 'ai-rule-builder-group-logic',
+  template: `
+    <div class="iot--rule-builder-header__dropdown">
+      <ibm-dropdown
+        theme="light"
+        [ngModel]="selected"
+        (ngModelChange)="selectedChange.emit($event)"
+        value="id"
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-group-logic.component.ts:11]: maybe document under the `anyAll` input that the displayed value will be `id`?

File: packages/angular/src/rule-builder/rule-builder-group-logic.component.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      >
+        <ibm-dropdown-list [items]="anyAll"></ibm-dropdown-list>
+      </ibm-dropdown>
+    </div>
+
+    <span>of the following are true</span>
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder-group-logic.component.ts:17]: Should this text be modifiable?

**Review by Donisius — CHANGES_REQUESTED** [review]

File: packages/angular/src/rule-builder/rule.component.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      <ng-template
+        *ngIf="hasTemplate()"
+        [ngTemplateOutlet]="getTemplate()"
+        [ngTemplateOutletContext]="{ $implicit: rule }"
+      >
+      </ng-template>
+      <div class="iot--rule-builder-rule__actions">
+        <button ibmButton="ghost" [iconOnly]="true" (click)="removeRule.emit(rule.id)">
+          <svg class="bx--btn__icon" ibmIcon="subtract" size="32"></svg>
+          <span class="bx--assistive-text">Remove rule</span>
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule.component.ts:53]: Is it too much to have this and the other assistive texts in this template be modifiable as well?

File: packages/angular/src/rule-builder/rule.component.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    <ng-container *ngIf="isRuleGroup">
+      <ai-rule-builder-group-logic [id]="rule.id" [(selected)]="rule.groupLogic">
+      </ai-rule-builder-group-logic>
+      <ng-container *ngFor="let r of rule.rules; let i = index">
+        <ai-rule
+          (addRule)="addRule.emit($event)"
+          (removeRule)="removeRule.emit($event)"
+          [columns]="columns"
+          [columnOperands]="columnOperands"
+          [(rule)]="rule.rules[i]"
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule.component.ts:16]: Can we just use `r`?

File: packages/angular/src/rule-builder/rule-builder.component.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        [(groupLogic)]="tree.groupLogic"
+        (addRule)="handleAddRule($event.id, $event.isGroup)"
+      ></ai-rule-builder-header>
+      <ng-container *ngFor="let rule of tree.rules; let i = index">
+        <ai-rule
+          (addRule)="handleAddRule($event.id, $event.isGroup)"
+          (removeRule)="handleRemoveRule($event)"
+          [columns]="columns"
+          [columnOperands]="columnOperands"
+          [(rule)]="tree.rules[i]"
```
Comment (Donisius) [inline packages/angular/src/rule-builder/rule-builder.component.ts:25]: can we just use `rule`?

**Review by Donisius — CHANGES_REQUESTED** [review]

File: packages/angular/src/rule-builder/rule.component.ts (hunk 8 above)
Comment (zvonimirfras) [reply packages/angular/src/rule-builder/rule.component.ts:16]: No, not for double binding. Angular.
Story works fine, build fails.

File: packages/angular/src/rule-builder/rule-builder.component.ts (hunk 9 above)
Comment (zvonimirfras) [reply packages/angular/src/rule-builder/rule-builder.component.ts:25]: No, not for double binding. Angular.
Story works fine, build fails. This is a workaround.

File: packages/angular/src/rule-builder/rule.component.ts (hunk 7 above)
Comment (zvonimirfras) [reply packages/angular/src/rule-builder/rule.component.ts:53]: Well, doh.

**Review by Donisius — APPROVED** [review]

**Review by davidicus — APPROVED** [review]

**Pull request merged**